### PR TITLE
Added support for specifying application name in config file and command line

### PR DESCRIPTION
**Pull request by siavashs**
This commit makes it possible to specify application name in configuration file:
```yaml
files:
  - path: locallog.txt
    application: my_app
```
or in command line:
```bash
$ remote_syslog my_app:locallog.txt
```

### CODE REVIEW

**Comment by siavashs** [comment]
@troy Thanks! :) Our organization has large number of databases and web servers. We recently started to use graylog as a hosted centralized log storage solution for syslogs but needed something to collect the log files and came across remote_syslog2. In order to create streams easily based on log categories we decided to use application name. We discussed about using symlinks but no one liked it so I implemented this feature.

@eric I can make it backward compatible by making the application part optional in command line and using a different key name in configuration file. Please let me know if you're interested to include the changes and I'll implement and add the changes to this pull request.

**Comment by aioue** [comment]
+1

**Comment by sebest** [comment]
@siavashs this is a really neat feature, @eric any update on merging this?

**Comment by troy** [comment]
@sebest the last comment from @siavashs is the current one; that is, @eric's comments about compatibility are still pending.

@siavashs, it looks like you may have asked a question that was blocking. The most important part is to continue using the filename as the sending value when none is given (on the CLI or in the config file). Beyond that, I propose 1 small change:

* call the attribute name `tag` (instead of `application`) to be consistent with the name that syslog uses

and 2 tests:

* add test with a wildcard, like this: `abc:/path/to/*.log`
* add test with multiple files and application/tag values in the config file

If you need any other guidance, by all means let us know. Thanks for contributing -- this is really close.

**Comment by siavashs** [comment]
Ok, I'll fix the conflicts and update the code as @troy suggested :)

**Comment by siavashs** [comment]
@troy I updated the code, and I made tags optional both in command line and config file.
The only breaking change is `path` in config file.

I was able to write a test for tags(not committed yet) but I'm not sure how to write the test for wildcard.

**Comment by troy** [comment]
@siavashs To clarify, this change is predicated on existing configs continuing to work. That is, this:

```
files:
  - locallog.txt
```

.. should continue to work. In other words, no existing users should need to touch their configs or CLI arguments. I'd guess 99% of people will continue using that, and the 1% of people who need this new feature can migrate to the new optional syntax. Does that make more sense? :)

If this PR doesn't break the existing config syntax, please feel free to correct me. From a cursory review and from your comment, it sounds like it does.

**Comment by troy** [comment]
Also, as part of preserving the existing config format as the standard, can you change `example_config.yml` back? I'm all for incorporating this into the readme though.

**Comment by siavashs** [comment]
@troy done! :)
The new configuration syntax is like command line now `tag:logfile_path`, and the tags are optional.

**Comment by troy** [comment]
Thanks, @siavashs!

Giving to @leonsodhi for first review. From a quick read, it looks like some code is duplicated and perhaps the default tag behavior might not be idiomatic.

**Comment by eric** [comment]
Would it be possible to provide a way to use both a "old" and "new" syntax so something like this would work?

``` yaml
files:
  - path: locallog.txt
    tag: local
  - production.log
  - path: railslog.txt
    tag: rails
```

which would prodice something like:

``` ruby
{"files"=>
  [{"path"=>"locallog.txt", "tag"=>"local"},
   "production.log",
   {"path"=>"railslog.txt", "tag"=>"rails"}]}
```

It would allow for people to continue using the old style of just having the values being a string, or a new style of the values being a map.

If we did allow for this, I expect as time went on, there could be more log-specific configuration that we would come across.

I understand that it could make the YAML parsing a little bit more complicated, because the fields would not have a one-to-one mapping to structs, but I think it would be worth it if it were possible.

**Comment by siavashs** [comment]
@eric that's what I wanted but the only solution that I found was using maps, like this: https://www.reddit.com/r/golang/comments/2zk0rw/uh_well_i_cannot_unmarshal_this_yaml_data/
I was afraid that using nested maps would make the code somehow complicated, but if you think it's a good idea I will do it :)

**Comment by eric** [comment]
Yeah, that yaml definition does look scary, but I think it's worth it for the usability.

**Comment by siavashs** [comment]
@eric that nested map did not work unless one added a `:` at the end of string.
Anyway I did it using interfaces and reflection :)
Sample configuration:
```yaml
files:
  - locallog.txt
  - path: locallog2.txt
    tag: foo
  - locallog3.txt
```
```
2015-10-13 15:50:14 INFO  remote_syslog.go:91 Forwarding locallog.txt
2015-10-13 15:50:14 INFO  remote_syslog.go:91 Forwarding locallog2.txt
2015-10-13 15:50:14 INFO  remote_syslog.go:91 Forwarding locallog3.txt
```

File: config_manager.go [hunk 1]
```diff
@@ -283,8 +297,26 @@ func (cm *ConfigManager) Poll() bool {
...
+			break
+		case reflect.Map:
+			m := v.Interface().(map[interface{}]interface{})
+			tag := reflect.ValueOf(m["tag"])
+			path := reflect.ValueOf(m["path"])
+			if tag.Kind() == reflect.String && path.Kind() == reflect.String {
+				logFiles = append(logFiles, LogFile{Tag: tag.String(), Path: path.String()})
+			}
+			break
+		default:
```
Comment (eric) [inline config_manager.go]: I feel like we should be reporting some sort of error here.

File: config_manager.go [hunk 2]
```diff
@@ -283,8 +297,26 @@ func (cm *ConfigManager) Poll() bool {
...
+		case reflect.String:
+			logFiles = append(logFiles, LogFile{Tag: "", Path: v.String()})
+			break
+		case reflect.Map:
+			m := v.Interface().(map[interface{}]interface{})
+			tag := reflect.ValueOf(m["tag"])
+			path := reflect.ValueOf(m["path"])
+			if tag.Kind() == reflect.String && path.Kind() == reflect.String {
+				logFiles = append(logFiles, LogFile{Tag: tag.String(), Path: path.String()})
+			}
```
Comment (eric) [inline config_manager.go]: It seems like we should be reporting some sort of error here in the `else`.

**Comment by eric** [comment]
:sparkles: this is so great.

With the error reporting that I referenced, I think this should be good to go.

**Comment by siavashs** [comment]
It now prints an error like this:
```
2015-10-13 19:27:09 ERROR  config_manager.go:317 Cloud not parse log file configuration: map[tag:1 path:test.txt]
```

**Comment by eric** [comment]
Sweet! Thanks!

**Comment by leonsodhi** [comment]
Thanks for implementing this, @siavashs. Just one thing from me at this point: could you tweak the error so that "Cloud" becomes "Could"?

**Comment by siavashs** [comment]
No problem, and sorry about the typo.

**Comment by siavashs** [comment]
@eric any update on merging this?

**Comment by troy** [comment]
@siavashs there's at least 1 set of bug fixes ahead of it (see [Pulls](https://github.com/papertrail/remote_syslog2/pulls)). It's on the radar, but is also a big enough change to warrant pre-release field testing. While I don't have an ETA, I'd guess weeks, not days nor months :)

**Comment by DannyBen** [comment]
What is the status of this PR? Are there plans to merge and release it? Or was it rejected?

In our configuration we must be able to specify the application for each log file, and I wish to avoid using the symlink approach.

**Comment by leonsodhi** [comment]
@DannyBen, there was one outstanding issue related to Windows systems that came up a while back. We ran through those internally today and I've included the details below.

@siavashs, I'm summarising what a number of PT folks have said, so feel free to ask for clarification if I'm slightly off. 

Because Windows uses a `:` character in file paths, could we remove the CLI support and just make this a config option?

**Comment by leonsodhi** [comment]
@siavashs feel free to ask us to make the change if you're short on time, although this is very much your PR - I don't want to steal your thunder.

**Comment by DannyBen** [comment]
FWIW, I very much like to see this feature merged in.

**Comment by siavashs** [comment]
@leonsodhi I will update this PR tomorrow to use = instead of : in command line.

**Comment by siavashs** [comment]
The command line now works like this:
```
tag=/path/to/log/file
tag=C:\path\to\log\file
```
I don't have a Windows machine at home for testing, but this should avoid any conflicts with Windows file system paths.

**Comment by leonsodhi** [comment]
Awesome! Thanks, @siavashs. I've just tested it here on a Windows and Linux box, and all looks good.  Seems like it's ready to be merged :-)

**Comment by DannyBen** [comment]
These are great news. Thanks for making it happen.

**Comment by siavashs** [comment]
No problem guys :)
BTW, I think it would be nice if we show the new CLI syntax in help output.

**Comment by DannyBen** [comment]
We are currently testing this for real world scenario and it seems to be working nicely, but a question / need for feature arises.

Previously, we used wildcards like this:
(the `{{ name }}` tag is from our ansible template, and holds the name of the currently deployed app)

```
  - path: /opt/rails/{{ name }}/log/*.log
```

Now, in order to utilize the new tag feature, we need to specify the logs one by one, like this:

```
  - path: /opt/rails/{{ name }}/log/app.log
    tag: {{ name }}_app
  - path: /opt/rails/{{ name }}/log/puma.log
    tag: {{ name }}_puma
  - path: /opt/rails/{{ name }}/log/crono.log
    tag: {{ name }}_crono
    ...
```

It could have been very helpful, if we could "backreference" the original tag (i.e. the filename without the extension) in someway, so we could still use wildcards like this:

```
  - path: /opt/rails/{{ name }}/log/*.log
    tag: tag: {{ name }}_%basename
```

Does this resonate with someone other than me?

**Comment by siavashs** [comment]
That is templating tag names with a combination of tag name and log file base name.
I'm not the one to decide about it but It should be easy to implement.

**Comment by DannyBen** [comment]
Another suggestion I want to put out there, is the ability to change the other tag - what is currently considered the hostname / system name.

Then, we will be able to configure like this:

```
- path: /opt/rails/{{ name }}/log/*.log
  system: {{ name }}
```

And the output will look like this:
( assuming `{{ name }}` evaluates to `rails01 and rails02` )
```
timestamp    rails01    crono     event details
timestamp    rails02    crono     event details
``` 
instead of
```
timestamp    rails    rails01_crono     event details
timestamp    rails    rails02_crono     event details
``` 

In other words - allow to set hostname per file, and not globally per actual physical machine.

**Comment by troy** [comment]
@DannyBen can you split that idea into a different issue (even if it's just what you wrote above, which is a good explanation of how it might work). Also, In the new issue, could you also add a bit about your own env and why you want a different sender for each file?

**Comment by DannyBen** [comment]
It will be not only my duty, but my pleasure. #126

**Comment by rmoriz** [comment]
docs missing…

**Pull request merged**